### DISCUSSION TITLE
next-swc's mysterious shakeExports option

### DISCUSSION THREAD

**Discussion by overlookmotel** [vercel/next.js#35487] [Help] [answered]
I wonder if anyone on NextJS team can give any background on the undocumented `shakeExports` option to `next-swc` which was [added in 12.0.8](https://github.com/vercel/next.js/pull/32253)?

I understand from #34010 that it's not intended to be exposed at present, and I'm not asking for it to be. I'm just wondering what the purpose of it is intended to be in the future? It appears to do much the same as the [`disableNextSsg` option](https://github.com/vercel/next.js/blob/da3760bbf5544077a6ffa513c84030d0c0672a69/packages/next/build/swc/options.js#L243) but with the list of exports to shake out configurable.

Reason I ask is that I've been playing with this option (via monkey-patching `next-swc-loader`) and it'd be very useful for something I'm working on. However, obviously it'd be unwise for me to rely on it if it may disappear at any minute.

I assume it's not been publicly exposed for a reason, and Next doesn't want to commit to maintaining the API at this point. But any background you could give for it's intended purpose would be really helpful in weighing up the risk of utilising it.

@padmaia you [requested](https://github.com/vercel/next.js/issues/31742) this option be added. Sorry to ping you out of the blue, but would you be able to shed any light?

**Comment by padmaia** [comment, marked as answer]
It is used internally by Next.js live. There are no plans to make it public at this point in time. It would be interesting to hear your use case though.

**Comment by overlookmotel** [comment]
Thanks very much for your reply @padmaia. Ah interesting.

If it's used in Next.js Live I guess I can assume it'll likely remain present in some form, even if the API changes.

My main use case is to define API in same file as page. You write the page like this:

```jsx
// pages/index.jsx
export default function Person({name}) {
  return <div>Hello {name}!</div>;
}

export async function get() {
  return { name: 'Bertrand' };
}

export async function post() {
  // Logic to update person's name
}
```

A loader I've added before SWC converts this to something like:

```jsx
// pages/index.jsx
export default function Page({name}) {
  return <div>Hello {name}!</div>;
}

export async function get() {
  return { name: 'Bertrand' };
}

export async function post() {
  // Logic to update person's name
}

// Added by my loader
export async function getServerSideProps(context) {
  return { props: await get(context) };
}
```

```js
// pages/api/index.js
import {get, post} from '../index.jsx';

export default async function(req, res) {
  const handler = req.method === 'POST' ? post : get,
    context = { params: req.query, query: req.query, req, res };
  res.status(200).json( await handler(context) );
}
```

Every page in my app has a corresponding API route and it's more natural to co-locate the data loading/mutation code in the page file along with the component, and not have to maintain a separate API page file for every route. This is just an extension of the idea of `getServerSideProps()` really, and quite similar to Remix's approach.

Where `shakeExports` comes in is for shaking out `export function get() {}` and `export function post() {}` from the client-side code, same as how NextJS already shakes out `getServerSideProps()`.

I could achieve the same thing with a Babel plugin, but keen to take advantage of SWC's speed.

I can think of other use cases where it's useful to export other functions / config from a page, which are used by another loader to modify the page in some way, but then should be deleted from the code before it gets Webpack-ed. e.g. implementing something similar to [babel-plugin-preval](https://www.npmjs.com/package/babel-plugin-preval).

If this isn't prying too much, can I ask what it's used for in Next.js Live?
